gnome-initial-setup/pages/site/gis-site-search-entry.c

```diff
@@ -0,0 +1,541 @@
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  switch (prop_id)
+    {
+      case PROP_SITES_AVAILABLE:
+        set_property_sites_available (GIS_SITE_SEARCH_ENTRY (object), value);
+        break;
+      case PROP_SITE:
+        gis_site_search_entry_set_site (GIS_SITE_SEARCH_ENTRY (object),
+                                        g_value_dup_object (value));
+        break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+        break;
+    }
+}
+
+static void
+get_property (GObject *object, guint prop_id,
+              GValue *value, GParamSpec *pspec)
+{
+  GisSiteSearchEntryPrivate *priv = gis_site_search_entry_get_instance_private (
+    GIS_SITE_SEARCH_ENTRY (object));
+
+  switch (prop_id)
+    {
+      case PROP_SITE:
+        g_value_set_boxed (value, priv->site);
+        break;
+      default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+        break;
+    }
+}
+
+static void
+entry_changed (GisSiteSearchEntry *entry)
+{
+  GisSiteSearchEntryPrivate *priv =
+      gis_site_search_entry_get_instance_private (entry);
+  GtkEntryCompletion *completion;
+  const gchar *text;
+
+  completion = gtk_entry_get_completion (GTK_ENTRY (entry));
+
+  gtk_entry_completion_set_match_func (gtk_entry_get_completion (GTK_ENTRY (entry)), matcher, NULL, NULL);
+  gtk_entry_completion_set_model (gtk_entry_get_completion (GTK_ENTRY (entry)), priv->model);
+
+  text = gtk_entry_get_text (GTK_ENTRY (entry));
+  if (!text || !*text)
+    set_site_internal (entry, priv->model, NULL, NULL);
+}
+
+static void
+string_append_safe (GString     *string,
+                    const gchar *field)
+{
+  if (field && field[0] != '\0')
+    g_string_append_printf (string, "%s%s", string->len > 0 ? ", " : "", field);
+}
+
+/* NOTE: this assumes the values provided here are already translated into the
+ * local language
+ */
+static gchar*
+get_site_display_name (GisSite *site)
+{
+  const gchar *id = gis_site_get_id (site);
+  const gchar *facility = gis_site_get_facility (site);
+  const gchar *street = gis_site_get_street (site);
+  const gchar *locality = gis_site_get_locality (site);
+  const gchar *region = gis_site_get_region (site);
+  const gchar *country = gis_site_get_country (site);
+
+  GString *string = g_string_new ("");
+
+  string_append_safe (string, id);
+  string_append_safe (string, facility);
+  string_append_safe (string, street);
+  string_append_safe (string, locality);
+  string_append_safe (string, region);
+  string_append_safe (string, country);
+
+  return g_string_free (string, FALSE);
+}
+
+static void
+set_site_internal (GisSiteSearchEntry *entry,
+                   GtkTreeModel       *model,
+                   GtkTreeIter        *iter,
+                   GisSite            *site)
+{
+  GisSiteSearchEntryPrivate *priv =
+      gis_site_search_entry_get_instance_private (entry);
+
+  if (priv->site)
```

Comment:
You can remove this check, `g_clear_object()` is designed so that you don't need to check before